File: hariom_sanpra/hariom_sanpra/page/hariom_material_alert/hariom_material_alert.py
import frappe
from frappe import _
from frappe.utils import cint, flt
from frappe.utils.nestedset import get_descendants_of


@frappe.whitelist()
def get_material_stock(company=None, warehouse=None, item_group=None, consumption_days=30):
	"""Return available stock and consumption cover grouped by Item Group and UOM."""
	if not frappe.has_permission("Item", "read") or not frappe.has_permission("Stock Ledger Entry", "read"):
		frappe.throw(_("Not permitted"), frappe.PermissionError)

	company = company or frappe.defaults.get_user_default("Company")
	consumption_days = max(cint(consumption_days), 1)
	conditions = ["item.disabled = 0"]
	values = {"consumption_days": consumption_days}

	if company:
		conditions.append("warehouse.company = %(company)s")
		values["company"] = company
	if warehouse:
		warehouses = [warehouse, *get_descendants_of("Warehouse", warehouse, ignore_permissions=True)]
		conditions.append("stock.warehouse IN %(warehouses)s")
		values["warehouses"] = warehouses
	raw_material_group = frappe.db.get_value(
		"Item Group", "Raw Material", ["lft", "rgt"], as_dict=True
	)
	if not raw_material_group:
		frappe.throw(_("Item Group Raw Material was not found"))

	item_group = item_group or "Raw Material"
	if item_group:
		group = frappe.db.get_value("Item Group", item_group, ["lft", "rgt"], as_dict=True)
		if not group:
			frappe.throw(_("Item Group {0} was not found").format(item_group))
		if group.lft < raw_material_group.lft or group.rgt > raw_material_group.rgt:
			frappe.throw(_("Please select Raw Material or one of its child Item Groups."))
		conditions.append(
			"item_group.lft >= %(group_lft)s AND item_group.rgt <= %(group_rgt)s"
		)
		values.update(group_lft=group.lft, group_rgt=group.rgt)

	rows = frappe.db.sql(
		f"""
		SELECT
			item.item_group,
			item.stock_uom,
			SUM(stock.actual_qty) AS actual_qty,
			SUM(COALESCE(bin.projected_qty, stock.actual_qty)) AS projected_qty,
			COUNT(DISTINCT stock.item_code) AS item_count,
			COALESCE(SUM(consumption.consumed_qty), 0) AS consumed_qty
		FROM (
			SELECT item_code, warehouse, SUM(actual_qty) AS actual_qty
			FROM `tabStock Ledger Entry`
			WHERE docstatus < 2
				AND is_cancelled = 0
				AND posting_date <= CURDATE()
			GROUP BY item_code, warehouse
		) stock
		INNER JOIN `tabItem` item ON item.name = stock.item_code
		INNER JOIN `tabItem Group` item_group ON item_group.name = item.item_group
		INNER JOIN `tabWarehouse` warehouse ON warehouse.name = stock.warehouse
		LEFT JOIN `tabBin` bin
			ON bin.item_code = stock.item_code AND bin.warehouse = stock.warehouse
		LEFT JOIN (
			SELECT item_code, warehouse, SUM(ABS(actual_qty)) AS consumed_qty
			FROM `tabStock Ledger Entry`
			WHERE docstatus < 2
				AND is_cancelled = 0
				AND actual_qty < 0
				AND posting_date >= DATE_SUB(CURDATE(), INTERVAL %(consumption_days)s DAY)
				AND posting_date <= CURDATE()
			GROUP BY item_code, warehouse
		) consumption
			ON consumption.item_code = stock.item_code AND consumption.warehouse = stock.warehouse
		WHERE {' AND '.join(conditions)}
		GROUP BY item.item_group, item.stock_uom
		HAVING ABS(actual_qty) > 0.000001 OR consumed_qty > 0.000001
		ORDER BY item.item_group, item.stock_uom
		""",
		values,
		as_dict=True,
	)

	for row in rows:
		row.actual_qty = flt(row.actual_qty)
		row.projected_qty = flt(row.projected_qty)
		row.consumed_qty = flt(row.consumed_qty)
		daily_consumption = row.consumed_qty / consumption_days
		row.days_left = max(row.actual_qty, 0) / daily_consumption if daily_consumption else None
		if row.actual_qty <= 0 or (row.days_left is not None and row.days_left <= 3):
			row.status = "danger"
		elif row.days_left is not None and row.days_left <= 10:
			row.status = "warning"
		else:
			row.status = "safe"

	return rows


@frappe.whitelist()
@frappe.validate_and_sanitize_search_inputs
def get_raw_material_item_groups(doctype, txt, searchfield, start, page_len, filters):
	"""Link-field options limited to Raw Material and all of its descendants."""
	if not frappe.has_permission("Item Group", "read"):
		frappe.throw(_("Not permitted"), frappe.PermissionError)

	raw_material_group = frappe.db.get_value(
		"Item Group", "Raw Material", ["lft", "rgt"], as_dict=True
	)
	if not raw_material_group:
		return []

	return frappe.db.sql(
		"""
		SELECT name
		FROM `tabItem Group`
		WHERE lft >= %(lft)s
			AND rgt <= %(rgt)s
			AND name LIKE %(txt)s
		ORDER BY lft
		LIMIT %(start)s, %(page_len)s
		""",
		{
			"lft": raw_material_group.lft,
			"rgt": raw_material_group.rgt,
			"txt": f"%{txt}%",
			"start": cint(start),
			"page_len": cint(page_len),
		},
	)

File: hariom_sanpra/hariom_sanpra/page/hariom_material_alert/hariom_material_alert.js
frappe.pages["hariom-material-alert"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Hariom Material Alert"),
		single_column: true,
	});

	new hariom_sanpra.MaterialAlert(page);
};

frappe.provide("hariom_sanpra");

hariom_sanpra.MaterialAlert = class MaterialAlert {
	constructor(page) {
		this.page = page;
		this.make_filter();
		this.make_body();
		this.page.set_primary_action(__("Refresh"), () => this.load(), "refresh");
		this.load();
	}

	make_filter() {
		this.company = this.page.add_field({
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_default("company"),
			change: () => {
				this.warehouse.set_value("");
				this.load();
			},
		});

		this.warehouse = this.page.add_field({
			fieldname: "warehouse",
			label: __("Warehouse"),
			fieldtype: "Link",
			options: "Warehouse",
			get_query: () => ({
				filters: {
					company: this.company.get_value(),
					disabled: 0,
				},
			}),
			change: () => this.load(),
		});

		this.item_group = this.page.add_field({
			fieldname: "item_group",
			label: __("Item Group"),
			fieldtype: "Link",
			options: "Item Group",
			default: "Raw Material",
			get_query: () => ({
				query: "hariom_sanpra.hariom_sanpra.page.hariom_material_alert.hariom_material_alert.get_raw_material_item_groups",
			}),
			change: () => this.load(),
		});
	}

	make_body() {
		this.$body = $(`
			<div class="material-alert-dashboard">
				<style>
					.material-alert-dashboard { padding: 8px 0 24px; }
					.material-alert-grid {
						display: grid;
						grid-template-columns: repeat(4, minmax(0, 1fr));
						gap: 28px;
					}
					.material-alert-card {
						position: relative;
						min-height: 178px;
						padding: 30px 36px;
						background: var(--card-bg, #fff);
						border: 1px solid var(--border-color, #e2e2e2);
						border-radius: 24px;
						box-shadow: 0 1px 2px rgba(0, 0, 0, 0.02);
					}
					.material-alert-title { font-size: 19px; font-weight: 600; line-height: 1.25; padding-right: 30px; }
					.material-alert-qty { margin-top: 22px; font-size: 27px; font-weight: 700; line-height: 1; }
					.material-alert-days { margin-top: 10px; font-size: 15px; line-height: 1.25; }
					.material-alert-dot { position: absolute; top: 34px; right: 34px; width: 18px; height: 18px; border-radius: 50%; }
					.status-safe .material-alert-dot { background: #08a317; }
					.status-safe .material-alert-days { color: #087514; }
					.status-warning .material-alert-dot { background: #ffb21a; }
					.status-warning .material-alert-days { color: #765000; }
					.status-danger .material-alert-dot { background: #dc3d3d; }
					.status-danger .material-alert-days { color: #8b1d1d; }
					@media (max-width: 991px) { .material-alert-grid { grid-template-columns: repeat(2, minmax(0, 1fr)); gap: 20px; } }
					@media (max-width: 600px) { .material-alert-grid { grid-template-columns: 1fr; gap: 16px; } .material-alert-card { min-height: 155px; padding: 25px 28px; } }
				</style>
				<div class="material-alert-result"></div>
			</div>
		`).appendTo(this.page.body);
		this.$result = this.$body.find(".material-alert-result");
	}

	load() {
		this.$result.html(`<div class="text-muted">${__("Loading stock...")}</div>`);
		frappe.call({
			method: "hariom_sanpra.hariom_sanpra.page.hariom_material_alert.hariom_material_alert.get_material_stock",
			args: {
				company: this.company.get_value(),
				warehouse: this.warehouse.get_value(),
				item_group: this.item_group.get_value() || "Raw Material",
				consumption_days: 30,
			},
			callback: (r) => this.render(r.message || []),
		});
	}

	render(rows) {
		if (!rows.length) {
			this.$result.html(`<div class="text-muted">${__("No stock found for the selected filters.")}</div>`);
			return;
		}

		const cards = rows.map((row) => {
			const title = frappe.utils.escape_html(row.item_group || __("Unassigned"));
			const uom = frappe.utils.escape_html(row.stock_uom || "");
			const qty = format_number(row.actual_qty, null, 2);
			let days_text = __("No recent consumption");
			if (row.days_left !== null && row.days_left !== undefined) {
				const days = Math.max(0, Math.floor(row.days_left));
				days_text = row.status === "danger"
					? __("{0} day(s) left · order now", [days])
					: __("{0} day(s) left", [days]);
			}

			return `
				<div class="material-alert-card status-${row.status || "safe"}">
					<span class="material-alert-dot" aria-hidden="true"></span>
					<div class="material-alert-title">${title}</div>
					<div class="material-alert-qty">${qty} ${uom}</div>
					<div class="material-alert-days">${days_text}</div>
				</div>`;
		}).join("");

		this.$result.html(`<div class="material-alert-grid">${cards}</div>`);
	}
};
